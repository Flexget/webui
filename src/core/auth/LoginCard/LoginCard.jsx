import React from 'react';
import PropTypes from 'prop-types';
import { reduxForm } from 'redux-form';
import CardActions from '@material-ui/core/CardActions';
import TextField from 'common/TextField';
import {
  Card,
  ErrorMessage,
  LoginButton,
  Content,
  LoginField,
} from './styles';

export const LoginCard = ({ handleSubmit, errorStatus }) => (
  <Card>
    <form onSubmit={handleSubmit}>
      <Content>
        <ErrorMessage>
          {errorStatus.message}
        </ErrorMessage>
        <LoginField
          name="username"
          component={TextField}
          id="username"
          label="Username"
          autoFocus
        />
        <LoginField
          name="password"
          component={TextField}
          id="password"
          label="Password"
          type="Password"
        />
      </Content>
      <CardActions>
        <LoginButton type="submit">
          Login
        </LoginButton>
      </CardActions>
    </form>
  </Card>
);

LoginCard.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  errorStatus: PropTypes.object,
};

LoginCard.defaultProps = {
  errorStatus: {},
};

export default reduxForm({
  form: 'login',
})(LoginCard);
