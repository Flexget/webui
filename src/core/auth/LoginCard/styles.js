import styled from '@emotion/styled';
import { Field } from 'redux-form';
import theme from 'theme';

import { Card as MUICard, CardContent, Button } from '@material-ui/core';

export const Card = styled(MUICard)`
  max-width: 40rem;
  margin: 0 1rem;
  ${theme.breakpoints.up('sm')} {
    margin: 0 auto;
  }
`;

export const ErrorMessage = styled.div`
  color: ${theme.palette.error[500]};
  text-align: center;
  padding: 1rem;
`;

export const LoginButton = styled(Button)`
  width: 100%;
`;

export const Content = styled(CardContent)`
  display: flex;
  flex-direction: column;
`;

export const LoginField = styled(Field)`
  padding-bottom: 1rem;
`;
