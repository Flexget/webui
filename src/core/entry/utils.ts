import { RawEntry, FieldNames, MappingType, GettersType, BaseEntry } from './types';
import { MovieEntry, movieFieldList, RawMovieEntry, MovieFieldNames } from './fields/movies';
import { SeriesEntry, seriesFieldList, SeriesFieldNames, RawSeriesEntry } from './fields/series';
import {
  EpisodeEntry,
  RawEpisodeEntry,
  episodesFieldList,
  EpisodeFieldNames,
} from './fields/episodes';

export const isMovie = (entry: RawEntry): entry is RawMovieEntry => !!entry.movieName;
export const isSeries = (entry: RawEntry): entry is RawSeriesEntry =>
  !!entry.seriesName && !entry.seriesSeason;
export const isEpisode = (entry: RawEntry): entry is RawEpisodeEntry & SeriesEntry =>
  !!(entry.seriesSeason && entry.seriesEpisode && entry.seriesName);

const getPropFn = <U>(mapping: MappingType<U>) => {
  const mapByField = mapping.reduce<Partial<Record<FieldNames<U>, string[]>>>(
    (res, item) =>
      Object.entries(item).reduce(
        (out, [k, v]) => ({
          ...out,
          [k]: [...(out[k] ?? []), v],
        }),
        res,
      ),
    {},
  );
  return (entry: U, field: FieldNames<U>) =>
    (mapByField[field] ?? ([] as string[])).reduce((res, val) => res ?? entry[val], undefined);
};

const makeGetters = <T>(names: Record<string, FieldNames<T>>, mapping: MappingType<T>) => {
  const getProp = getPropFn<T>(mapping);
  return (entry: T) =>
    Object.values(names).reduce(
      (obj, key: FieldNames<T>) => ({
        ...obj,
        get [key]() {
          return getProp(entry, key);
        },
      }),
      {} as GettersType<T>,
    );
};

const getMovieProps = makeGetters<RawMovieEntry>(MovieFieldNames, movieFieldList);
const getSeriesProps = makeGetters<RawSeriesEntry>(SeriesFieldNames, seriesFieldList);
const getEpisodeProps = makeGetters<RawEpisodeEntry>(EpisodeFieldNames, episodesFieldList);

export const toMovieEntry = (entry: RawMovieEntry): MovieEntry => {
  return {
    ...entry,
    ...getMovieProps(entry),
  };
};

export const toSeriesEntry = (entry: RawSeriesEntry): SeriesEntry => ({
  ...entry,
  ...getSeriesProps(entry),
});

export const toEpisodeEntry = (entry: RawEpisodeEntry): EpisodeEntry => ({
  ...entry,
  ...getEpisodeProps(entry),
});

export const toDefaultEntry = (entry: BaseEntry): BaseEntry => ({
  ...entry,
});
