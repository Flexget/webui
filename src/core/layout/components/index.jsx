import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Logo from 'core/layout/components/Logo';
import Navbar from 'core/layout/components/Navbar';
import SideNav from 'core/layout/components/SideNav';
import LoadingBar from 'components/LoadingBar';
import ErrorStatus from 'components/ErrorStatus';
import InfoStatus from 'components/InfoStatus';
import {
  Wrapper,
  Main,
  Header,
  LogoWrapper,
  Nav,
  SideBar,
  Content,
} from './styles';

class Layout extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
  };

  state = {
    sideBarOpen: (window.matchMedia && !!window.matchMedia('(min-width: 600px)').matches) || false,
  };

  toggleSideBar = () => {
    this.setState(({ sideBarOpen }) => ({
      sideBarOpen: !sideBarOpen,
    }));
  }

  render() {
    const { children } = this.props;
    const { sideBarOpen } = this.state;

    return (
      <Wrapper>
        <Header>
          <LogoWrapper>
            <Logo sideBarOpen={sideBarOpen} />
          </LogoWrapper>
          <Nav>
            <Navbar toggle={this.toggleSideBar} />
            <LoadingBar />
          </Nav>
        </Header>
        <Main>
          <SideBar>
            <SideNav
              sideBarOpen={sideBarOpen}
              toggle={this.toggleSideBar}
            />
          </SideBar>
          <Content open={sideBarOpen}>
            { children }
          </Content>
          <ErrorStatus />
          <InfoStatus />
        </Main>
      </Wrapper>
    );
  }
}

export default Layout;
