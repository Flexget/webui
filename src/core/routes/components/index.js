import { connect } from 'react-redux';
import Routes from './Routes';

function mapStateToProps({ routes }) {
  return {
    routes: Object.values(routes),
  };
}

export default connect(mapStateToProps)(Routes);
