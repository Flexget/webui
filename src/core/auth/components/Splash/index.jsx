import React from 'react';
import WaveSpinner from 'components/WaveSpinner';
import { Logo } from './styles';

const SplashScreen = () => (
  <div>
    <Logo />
    <WaveSpinner />
  </div>
);


export default SplashScreen;
