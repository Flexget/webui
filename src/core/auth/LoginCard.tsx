import React, { FC } from 'react';
import { Form, Formik } from 'formik';
import { useContainer } from 'unstated-next';
import { css } from '@emotion/core';
import styled from '@emotion/styled';
import { CardContent, Card, Button, CardActions } from '@material-ui/core';
import { useFlexgetAPI } from 'core/api';
import { Method } from 'utils/fetch';
import TextField from 'common/inputs/formik/TextField';
import theme from 'core/theme';
import { AuthContainer } from './container';
import { LoginReq } from './types';

const card = css`
  max-width: 40rem;
  margin: 0 1rem;
  ${theme.breakpoints.up('sm')} {
    margin: 0 auto;
  }
`;

const field = css`
  padding-bottom: 1rem;
`;

const ErrorMessage = styled.div`
  color: ${theme.palette.error[500]};
  text-align: center;
  padding: 1rem;
`;

const content = css`
  display: flex;
  flex-direction: column;
`;

const LoginCard: FC = () => {
  const initialValues: LoginReq = { username: 'flexget', password: '' };
  const [, setLoggedIn] = useContainer(AuthContainer);

  const [{ error, loading }, login] = useFlexgetAPI('/auth/login', Method.Post);

  return (
    <Card css={card}>
      <Formik
        initialValues={initialValues}
        onSubmit={async (values, actions) => {
          const resp = await login(values);
          if (resp.ok) {
            actions.resetForm();
            setLoggedIn(true);
          }
        }}
      >
        <Form>
          <CardContent css={content}>
            <ErrorMessage>{error?.message}</ErrorMessage>
            <TextField css={field} name="username" id="username" label="Username" autoFocus />
            <TextField css={field} name="password" id="password" label="Password" type="Password" />
          </CardContent>
          <CardActions>
            <Button type="submit" fullWidth disabled={loading}>
              Login
            </Button>
          </CardActions>
        </Form>
      </Formik>
    </Card>
  );
};

export default LoginCard;
