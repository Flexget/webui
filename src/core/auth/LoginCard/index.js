import { connect } from 'react-redux';
import actions from 'core/auth/state/actions';
import LoginCard from './LoginCard';

export function mapStateToProps({ status }) {
  return {
    initialValues: {
      username: 'flexget',
    },
    errorStatus: status.error || {},
  };
}

function mapDispatchToProps(dispatch) {
  return {
    onSubmit: ({ username, password }) => dispatch(actions.login.request(username, password)),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(LoginCard);
export { LoginCard };
