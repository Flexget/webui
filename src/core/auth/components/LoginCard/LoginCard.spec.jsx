import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import LoginCardFull, { LoginCard } from 'core/auth/components/LoginCard/LoginCard';
import { mapStateToProps } from 'core/auth/components/LoginCard';
import { themed, provider } from 'utils/tests';

describe('core/Login/LoginCard', () => {
  describe('LoginCard', () => {
    it('renders correctly', () => {
      const tree = renderer.create(provider(themed(<LoginCardFull />))).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should call handleSubmit on submit', () => {
      const handleSubmit = jest.fn();
      const wrapper = shallow(<LoginCard
        handleSubmit={handleSubmit}
        classes={{}}
      />);

      wrapper.find('form').simulate('submit');
      expect(handleSubmit).toHaveBeenCalled();
    });
  });


  describe('mapStateToProps', () => {
    it('should return no errors if there are none', () => {
      expect(mapStateToProps({ status: { } })).toMatchSnapshot();
    });

    it('should return errors if there are errors', () => {
      expect(mapStateToProps({ status: { error: { message: 'Invalid Credentials' } } })).toMatchSnapshot();
    });
  });
});
