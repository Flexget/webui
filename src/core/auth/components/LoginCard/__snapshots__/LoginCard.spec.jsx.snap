// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`core/Login/LoginCard LoginCard renders correctly 1`] = `
.glamor-5 {
  max-width: 40rem;
  margin: 0 1rem;
}

.glamor-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.glamor-0 {
  text-align: center;
  padding: 1rem;
}

.glamor-1 {
  padding-bottom: 1rem;
}

.glamor-4 {
  width: 100%;
}

@media (min-width:600px) {
  .glamor-5 {
    margin: 0 auto;
  }
}

<div
  className="MuiPaper-root-1 MuiPaper-shadow2-5 MuiPaper-rounded-2 glamor-5 egcdk620"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="MuiCardContent-root-28 glamor-3 egcdk623"
    >
      <div
        className="glamor-0 egcdk621"
      />
      <div
        aria-describedby={undefined}
        className="MuiFormControl-root-29 glamor-1 egcdk624"
        onBlur={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
      >
        <label
          className="MuiFormLabel-root-39 MuiInputLabel-root-33 MuiInputLabel-formControl-34 MuiInputLabel-animated-37"
          data-shrink={false}
          htmlFor="username"
        >
          Username
        </label>
        <div
          className="MuiInput-root-43 MuiInput-formControl-44 MuiInput-inkbar-45 MuiInput-underline-49"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <input
            aria-invalid={false}
            aria-required={false}
            autoComplete={undefined}
            autoFocus={true}
            className="MuiInput-input-52"
            defaultValue={undefined}
            disabled={false}
            id="username"
            name="username"
            onChange={[Function]}
            onKeyDown={undefined}
            onKeyUp={undefined}
            placeholder={undefined}
            readOnly={undefined}
            required={undefined}
            rows={undefined}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        aria-describedby={undefined}
        className="MuiFormControl-root-29 glamor-1 egcdk624"
        onBlur={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
      >
        <label
          className="MuiFormLabel-root-39 MuiInputLabel-root-33 MuiInputLabel-formControl-34 MuiInputLabel-animated-37"
          data-shrink={false}
          htmlFor="password"
        >
          Password
        </label>
        <div
          className="MuiInput-root-43 MuiInput-formControl-44 MuiInput-inkbar-45 MuiInput-underline-49"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <input
            aria-invalid={false}
            aria-required={false}
            autoComplete={undefined}
            autoFocus={undefined}
            className="MuiInput-input-52 MuiInput-inputType-55"
            defaultValue={undefined}
            disabled={false}
            id="password"
            name="password"
            onChange={[Function]}
            onKeyDown={undefined}
            onKeyUp={undefined}
            placeholder={undefined}
            readOnly={undefined}
            required={undefined}
            rows={undefined}
            type="Password"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="MuiCardActions-root-58"
    >
      <button
        className="MuiButtonBase-root-75 MuiButton-root-60 MuiCardActions-action-59 glamor-4 egcdk622"
        disabled={false}
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        <span
          className="MuiButton-label-61"
        >
          Login
        </span>
        <span
          className="MuiTouchRipple-root-77"
        />
      </button>
    </div>
  </form>
</div>
`;

exports[`core/Login/LoginCard mapStateToProps should return errors if there are errors 1`] = `
Object {
  "errorStatus": Object {
    "message": "Invalid Credentials",
  },
  "initialValues": Object {
    "username": "flexget",
  },
}
`;

exports[`core/Login/LoginCard mapStateToProps should return no errors if there are none 1`] = `
Object {
  "errorStatus": Object {},
  "initialValues": Object {
    "username": "flexget",
  },
}
`;
